search-handlers/src/main/java/no/unit/nva/search/ExportResourceHandler.java

```diff
@@ -75,12 +75,61 @@ public ExportResourceHandler(
   public String processS3Input(Void input, RequestInfo requestInfo, Context context)
       throws BadRequestException {
 
-    return RecursiveScrollQuery.builder()
-        .withInitialResponse(initialResponse)
-        .withScrollTime(SCROLL_TTL)
-        .build()
-        .doSearch(scrollClient)
-        .toCsvText();
+    var currentPageSize = INITIAL_HITS_PER_PAGE;
+    AttemptResponse response;
+    do {
+      response = attemptsWithPageSize(currentPageSize, requestInfo);
+      if (SIZE_LIMIT_EXCEEDED.equals(response.status)) {
+        var nextPageSize = currentPageSize / TWO;
+        LOGGER.info(
+            "Request entity too large encountered with page size {}, trying again with {}",
+            currentPageSize,
+            nextPageSize);
+        currentPageSize = nextPageSize;
+      } else if (OTHER_FAILURE.equals(response.status)) {
+        throw new RuntimeException(response.causeOfFailure);
+      }
+    } while (SIZE_LIMIT_EXCEEDED.equals(response.status));
+
+    return response.result;
+  }
+
+  AttemptResponse attemptsWithPageSize(int pageSize, RequestInfo requestInfo)
+      throws BadRequestException {
+    try {
+      var initialResponse =
+          ResourceSearchQuery.builder()
+              .fromRequestInfo(requestInfo)
+              .withParameter(FROM, ZERO)
+              .withParameter(SIZE, Integer.toString(pageSize))
+              .withParameter(AGGREGATION, NONE)
+              .withParameter(NODES_INCLUDED, INCLUDED_NODES)
+              .withRequiredParameters(SORT)
+              .build()
+              .withFilter()
+              .requiredStatus(PUBLISHED, PUBLISHED_METADATA)
+              .apply()
+              .withScrollTime(SCROLL_TTL)
+              .doSearch(opensearchClient, Words.RESOURCES)
+              .swsResponse();
+
+      return AttemptResponse.success(
+          RecursiveScrollQuery.builder()
+              .withInitialResponse(initialResponse)
+              .withScrollTime(SCROLL_TTL)
+              .build()
+              .doSearch(scrollClient, Words.RESOURCES)
+              .toCsvText());
+    } catch (CompletionException completionException) {
+      if (completionException.getCause() instanceof OpenSearchClientException
```

Comment:
x instanceof OSC osc && osc.getCause…

Reply:
Used instanceof with expression and extracted a method to make it easier to read and understand.